js/src/components/paid-ads/convertToAssetGroupUpdateBody.js

```diff
@@ -0,0 +1,91 @@
+/**
+ * Internal dependencies
+ */
+import { ASSET_KEY, ASSET_FORM_KEY } from '.~/constants';
+
+/**
+ * @typedef {import('.~/components/types.js').AssetGroupFormValues} AssetGroupFormValues
+ * @typedef {import('.~/data/types.js').AssetEntityGroup} AssetEntityGroup
+ * @typedef {import('.~/data/types.js').AssetOperations} AssetOperations
+ * @typedef {import('.~/data/types.js').AssetEntityGroupUpdateBody} AssetEntityGroupUpdateBody
+ */
+
+/**
+ * Differences the assets form values against the existing asset entity group and
+ * returns the updating operations of the asset group.
+ *
+ * @param  {AssetEntityGroup} assetGroup The existing asset entity group as the base data for comparing the differences.
+ * @param  {AssetGroupFormValues} values The assets form values to be compared.
+ * @return {AssetOperations[]} The creation and deletion operations for updating the asset group.
+ */
+export function diffAssetOperations( assetGroup, values ) {
+	const operations = [];
+	const assets = assetGroup.assets;
+
+	function pushDeletions( key, ...assetEntities ) {
+		const deletions = assetEntities.map( ( assetEntity ) => ( {
+			...assetEntity,
+			// Setting the `content` to null indicates the asset deletion operation.
+			content: null,
+			field_type: key,
+		} ) );
+		operations.push( ...deletions );
+	}
+
+	Object.values( ASSET_KEY ).forEach( ( key ) => {
+		const contents = [ values[ key ] ].flat().filter( Boolean );
+		const assetEntities = [ assets[ key ] ].flat().filter( Boolean );
+		let entityIndex = 0;
+
+		contents.forEach( ( content ) => {
```

Comment:
I as having a little bit of a trouble understanding this function. In concrete when we should push a deletion operation and when a creation.

Reply:
Added code comment in https://github.com/woocommerce/google-listings-and-ads/pull/1874/commits/50bffc816f65f5b03fe13ae553b96be9c92a5bb6. Do let me know if the explanation needs adjustments.

Reply:
I was playing a bit with the new version. And considering all the previous comments and discussions. Seems like the only part we discuss is about this function. 

the new comments are nice, they clarified a bunch. However I think it would be more useful to add the steps in line with the code to facilitate understanding for the code newcomers in the future.

```
/**
 * Differences the assets form values against the existing asset entity group and
 * returns the updating operations of the asset group.
 *
 * @param  {AssetEntityGroup} currentAssetEntityGroup The existing asset entity group as the base data for comparing the differences.
 * @param  {AssetGroupFormValues} formValues The assets form values to be compared.
 * @return {AssetOperations[]} The creation and deletion operations for updating the asset group.
 */
export function diffAssetOperations( currentAssetEntityGroup, formValues ) {
	const operations = [];

	function pushDeletions( key, ...assetEntities ) {
		const deletions = assetEntities.map( ( assetEntity ) => ( {
			...assetEntity,
			// Setting the `content` to null indicates the asset deletion operation.
			content: null,
			field_type: key,
		} ) );
		operations.push( ...deletions );
	}

	// The assets of Google Ads are immutable once created. Therefore if it wants to update
	// a content of an existing asset, the processing will be made by two operations:
	// 1. Delete that existing asset.
	// 2. Create a new asset.
	//
	// In the general case, the creation order of the contents in an asset is also the order
	// of the asset's contents after fetching the asset group. Thus, this loop compares the
	// contents of form values and the contents of the assets in the existing entity group by
	// the same key and one by one in order, and then differences out the operations.

	// Loops all possible asset keys one by one
	Object.values( ASSET_KEY ).forEach( ( key ) => {
		const currentKeyFormContents = [ formValues[ key ] ]
			.flat()
			.filter( Boolean );
		const currentKeyAssetEntities = [
			currentAssetEntityGroup.assets[ key ],
		]
			.flat()
			.filter( Boolean );
		let entityIndex = 0;

		// For the current Assets Key. Compares the contents of the form values and the contents current assets for that key.
		currentKeyFormContents.forEach( ( content ) => {
			do {
				const assetEntity = currentKeyAssetEntities[ entityIndex ];

				// If a content of the form value is the same as the current asset content and with same index position.
				// this means there is no  operation needed and we should continue with the next form asset content.
				if ( content === assetEntity?.content ) {
					break;
				}

				// it it's different order or content, we should delete that assetEntity
				if ( assetEntity ) {
					pushDeletions( key, assetEntity );
				}

				entityIndex += 1;
			} while ( entityIndex < currentKeyAssetEntities.length );

			// if at this point entityIndex is bigger than the current asset entities it means there is a new form content asset to be added as entity
			if ( entityIndex >= currentKeyAssetEntities.length ) {
				operations.push( {
					// Setting the `id` to null indicates the asset creation operation.
					id: null,
					content,
					field_type: key,
				} );
			}

			entityIndex += 1;
		} );

		// At this point. If an asset content is not found in the contents of the form value, this means it's a deletion
		pushDeletions( key, ...currentKeyAssetEntities.slice( entityIndex ) );
	} );

	return operations;
}
```

p.s The good thing is that after discussion and changes I now understand fully (I guess) the logic of this function. I found really a gotcha the fact that the order of the assets is relevant

Reply:
Thanks for the suggestion. Added a few more code comments in https://github.com/woocommerce/google-listings-and-ads/pull/1874/commits/c0662c7474916968a8952f14bf61f7e4c4b7e18e. I kept the logic bullets because they describe the concept but do not map to the actual lines of code one by one.